### ISSUE TITLE
DTLS failure causes losing notifications

### ISSUE THREAD

**Issue by kiemlicz** [OpenMobileAlliance/OMA_LwM2M_for_Developers#161]
Given following information:

(from CoAP Observe RFC)
- "All notifications resulting from a GET request with an Observe Option MUST be returned within the same epoch of the same connection as the request."
- "An acknowledgement message signals to the server that the client is alive and interested in receiving further notifications; if the server does not receive an acknowledgement in reply to a confirmable notification, it will assume that the client is no longer interested and will eventually remove the associated entry from the list of observers" In this example, the LwM2M client is observation server.

In case of losing the DTLS state, (according to quoted RFCs) established observations are no longer valid.
The LwM2M server won't see any notifications (which will be ok if he didn't specify pmin and pmax), LwM2M client will drop the observation.
Is it even possible that LWM2M server will know that something went wrong?

**Comment by boaks** [comment]
issues:
#137 
#71 

Simply: if you lose the DTLS state, you must use `register` (not `update`)

**Comment by dnav** [comment]
This issue will be addressed in the next TS.
A new DTLS session will explicitely require a new registration and thus new observation requests.

Regards,

**Comment by sbernard31** [comment]
What about session resuming ? this will not be a new session so following the last OMA-TS-LightweightM2M-V1_0-20161031-D TS no need to do a new registration but as the coap spec say same epoch and same connection. Observation will not works anymore :/

The real problem is the constraint in CoAP is too strong :
"All notifications resulting from a GET request with an Observe Option MUST be returned within the same epoch of the same connection as the request."

If we want to make it works keeping it simple, we should change this at CoAP side instead of just adding coupling between all concept.

From my point of view, the CoAP constraint should be "response of request should be returned by the same DTLS identity or same session if there is no authentication" as daniel Maier explain [here](https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues/137#issuecomment-249915062).

As in LWM2M there is authentication, DTLS session lifecycle will not be linked to registration or observation lifecyle.

**Comment by boaks** [comment]
`In case of losing the DTLS state`

May be this mean different things, but in may answer I assumed, that the state (including the session ID) was lost and therefore I don't see, how to resume.

**Comment by sbernard31** [comment]
You're right :). So the problem I raised before is not directly linked to this issue but it is still there.
And this issue(#161) would be fixed if the CoAP constraint was relaxed, as I proposed.

**Comment by kiemlicz** [comment]
Yes, I was not precise enough, sorry about that.
By "losing the DTLS state" we mean:
 1. Both: seqNo & epoch is gone but SessionTicket is available
 2. Whole DTLS session is gone (nothing is available)

Ad 1. This one is the interesting one. The re-registration seems to be an overkill as the server could resume session. But it is unclear if LwM2M server can send _ClientHello_ (_HelloRequest_ maybe?). Moreover the CoAP constraints are too tight and notifications will be lost either way (as epoch & seqNo resets).
Ad 2. Based on @boaks reply, client needs to react to DTLS failures and re-register.

> The real problem is the constraint in CoAP is too strong

> If we want to make it works keeping it simple, we should change this at CoAP side instead of just adding coupling between all concept.

This is music to my ears. 
After all why such constraint do exist in CoAP? Isn't it also an overkill?

To sum up:
 - if the DTLS session doesn't exist (not even SessionTicket) - LwM2M client must register. Spec needs to be updated
 - if the DTLS can be resumed, CoAP spec needs to be relaxed (any chance of succeeding?).

**Comment by sbernard31** [comment]
>After all why such constraint do exist in CoAP? Isn't it also an overkill?
This question was asked several time and nobody seems really able to answer...
Currently, this problem is known at IETF CoRE working group and some people seems ok to say this constraint is too strong.

**Comment by hannestschofenig** [comment]
The underlying problem is that the record layer in DTLS requires information from the transport layer to determine which security context to select. This problem is described in Section 4.1 of https://tools.ietf.org/html/draft-fossati-tls-iot-optimizations-00.

What the CoAP Observe draft tried to do is to come up with a clever way to use some additional information, which utilizes the epoch field as an additional identifier. Unfortunately, it does not quite work since the epoch serves a completely different purpose and it starts from zero. Additionally, the epoch value changes over time when you rekey the connection. 

The solution to the problem is described in Section 4.2 of https://tools.ietf.org/html/draft-fossati-tls-iot-optimizations-00.

**Comment by sbernard31** [comment]
For me https://tools.ietf.org/html/draft-fossati-tls-iot-optimizations-00 fixed a DTLS issue but not the CoAP one. This is 2 different issues.

See discussion [here](https://mailarchive.ietf.org/arch/search/?email_list=core&gbt=1&index=0vwuEnq6IRHJc70l3BTggX6p9vs).

I think sentence below should be removed from the spec  :
"When a new DTLS Session is started, or in NoSec mode when the LWM2M Client IP address changes, the Client MUST register again to the LWM2M Server."

We should not couple this concept. For security purpose the TLS spec advice to define session lifetime to 24h. With this constraint, thats means that most of the time a registration will live only for 24h ... or Users will extend TLS session lifetime  for a bad reason ....

**Comment by hannestschofenig** [comment]
Short update: At the IETF#99 meeting the TLS WG agreed to work on a connection ID concept for DTLS 1.2 and 1.3. I hope that there will be some progress by the next IETF meeting on this topic.

Regarding the mentioned sentence in the LwM2M spec: 
"When a new DTLS Session is started, or in NoSec mode when the LWM2M Client IP address changes, the Client MUST register again to the LWM2M Server."

We will discuss adjusting the sentence given the new developments around the connection id (in the IETF) and the CoAP over TCP addition (in LwM2M v1.1).

**Comment by boaks** [comment]
Starting with your (@hannestschofenig) e-mail (see https://www.ietf.org/mail-archive/web/core/current/msg07813.html) there was a long, still ongoing, discussion about the request/response matching especially, if a notify-response extends the timespan into uncertainty.
 
FMPOV https://www.ietf.org/mail-archive/web/core/current/msg08821.html opens the way to find better matching specifications to ensure, that endpoints have the right identity. 
For me this includes lower layer matching (e.g. DTLS based identity using the credentials, if their unique) and upper layer matching (e.g. using LWM2M registration / update), both independent from a potential unstable IPv4 address. 
Californium is currently preparing to offer "none RFC7252 conform" matching possibilities of notify-responses.

But I'm also looking forward to the DTLS1.3 and connection ID work, but I'm not sure, if the "hash-chained" connection ID will work in a clustered environment. If we have to accept, that many clients have to establish a new DTLS session much more frequently, the notify response matching must be relaxed much more.

**Comment by hannestschofenig** [comment]
There is work in the IETF on the DTLS Connection ID in progress, see https://tools.ietf.org/html/draft-rescorla-tls-dtls-connection-id-01. 

As part of LwM2M v1.1 we will re-evaluate the relationship between the transport and the LwM2M messaging. This may require new work in the CORE working group on Observe and CoAP itself.
